### PR TITLE
ROX-20334: compliance stats by profile

### PR DESCRIPTION
**Pull request by dashrews78**
## Description

Updates to group stats by profile.  This is similar to what is used for the card groupings by scan schedule currently.  

Due to the way the search framework joins tables and the fact that each cluster has a copy of a result and a profile and a scan, I added a `profile_ref_id` and `scan_ref_id` respectively.  These are UUIDs deterministaclly created from the `cluster_id` and bits of the profile or scan depending on which is generated.  These are created and written when the data is ingested and greatly simplifies and improves our ability to join across these 3 objects.

This will need a migration.  Tracked with https://issues.redhat.com/browse/ROX-23134

## Checklist
- [ ] Investigated and inspected CI test results
- [ ] Unit test and regression tests added
- [ ] Evaluated and added CHANGELOG entry if required
~- [ ] Determined and documented upgrade steps~
~- [ ] Documented user facing changes (create PR based on [openshift/openshift-docs]~(https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))

If any of these don't apply, please comment below.

### Here I tell how I validated my change

Spun up 2 clusters and ran a few different compliance scans then used the api to get the results.  Executed the following API queries along with the results:

```
# profile stats
curl --insecure "https://central-stackrox.apps.ds-03-14-sedate-smash-correl.ocp.infra.rox.systems:443/v2/compliance/scan/stats/profiles" -u admin:XXX | json_pp
```
resutls - 
[all.json](https://github.com/stackrox/stackrox/files/14606115/all.json)

```
# profile stats query by profile
curl --insecure "https://central-stackrox.apps.ds-03-14-sedate-smash-correl.ocp.infra.rox.systems:443/v2/compliance/scan/stats/profile/ocp4-e8" -u admin:XXX | json_pp
```
results - 
[e8_only.json](https://github.com/stackrox/stackrox/files/14606117/e8_only.json)

```
# profile stats query by profile and cluster
curl --insecure "https://central-stackrox.apps.ds-03-14-sedate-smash-correl.ocp.infra.rox.systems:443/v2/compliance/scan/stats/profile?query=Compliance%20Profile%20Name:ocp4-e8%2BCluster%20ID%3Ac72e8af3-8179-4879-8e23-cb9dd558a918" -u admin:XXX | json_pp
```
results - 
[e8_remote.json](https://github.com/stackrox/stackrox/files/14606123/e8_remote.json)

### Reminder for reviewers

In addition to reviewing code here, reviewers **must** also review testing and request further testing in case the
performed one does not seem sufficient. As a reviewer, you must not approve the change until you understand the
performed testing and you are satisfied with it.

### CODE REVIEW

File: central/complianceoperator/v2/checkresults/datastore/datastore_impl.go [hunk 1]
```diff
@@ -113,6 +113,44 @@ func (d *datastoreImpl) ComplianceCheckResultStats(ctx context.Context, query *v
...
+func (d *datastoreImpl) ComplianceProfileResultStats(ctx context.Context, query *v1.Query) ([]*ResourceResultCountByProfile, error) {
+	defer metrics.SetDatastoreFunctionDuration(time.Now(), "ComplianceOperatorCheckResultV2", "ComplianceProfileResultStats")
+
+	var err error
+	query, err = withSACFilter(ctx, resources.Compliance, query)
+	if err != nil {
+		return nil, err
+	}
+
+	cloned := query.Clone()
```
Comment (SimonBaeumer) [inline central/complianceoperator/v2/checkresults/datastore/datastore_impl.go:126]: For my understanding: Why do you need to clone the query here?

File: central/complianceoperator/v2/checkresults/service/service_impl.go [hunk 2]
```diff
@@ -98,9 +98,24 @@ func (s *serviceImpl) GetComplianceScanResults(ctx context.Context, query *v2.Ra
...
+	if err != nil {
+		return nil, errors.Wrapf(errox.InvalidArgs, "Unable to parse query %v", err)
+	}
+
+	// Fill in pagination.
+	paginated.FillPaginationV2(parsedQuery, query.GetPagination(), maxPaginationLimit)
+
+	scanResults, err := s.complianceResultsDS.ComplianceProfileResultStats(ctx, parsedQuery)
+	if err != nil {
+		return nil, errors.Wrapf(err, "Unable to retrieve compliance profile scan stats for %v", query)
```
Comment (SimonBaeumer) [inline central/complianceoperator/v2/checkresults/service/service_impl.go:113]: I don't know how it is done in other error logs, but adding + prints more info. If it is not added in other logs and we didn't had problems just dismiss this comment.

```suggestion
		return nil, errors.Wrapf(err, "Unable to retrieve compliance profile scan stats for %+v", query)
```

File: central/convert/internaltov2storage/utils.go [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+	"github.com/stackrox/rox/pkg/logging"
+	"github.com/stackrox/rox/pkg/uuid"
+)
+
+var (
+	log = logging.LoggerForModule()
+)
+
+// BuildProfileRefID returns a UUID based off cluster, profileID, and product type
+func BuildProfileRefID(clusterID string, profileID string, productType string) string {
```
Comment (SimonBaeumer) [inline central/convert/internaltov2storage/utils.go:13]: I can see how this simplifies doing joins but carries the same disadvantages as a cache.
That means, when a some of the data in `BuildProfileRefID` needs to change this link returns another result, that would mean in case of a UPDATE operations this is not deterministically anymore.

I've tried to enhance the documentation that it is clear why a special `BuildProfileRefID` is generated.
```
// BuildProfileRefID returns a UUID based off cluster, profileID, and product type to simplify SQL joins in SELECT statements.
```

Additionally we should forbid SQL updates on result fields to `clusterID`, `profileID` and product type that these fields are **not writable** anymore, instead if a user/dev wants to update these fields they need to delete all the respective results and let Central re-create them.

While this is clear to us now while implementing compliance v2, in longer-terms this code might be hard to understand because it solves a specific problem and introduces an implicit dependency on the `BuildProfileRefID` function.

TL;DR, I think we need:

- Slightly better documentation on this function to explain what it achieves and when someone should use it
- Prevent updates on `profileID`, `clusterID` and `product type`. (Whereas I see product type as critical when a name changes, e.g. RHACS to StackRox or OCP to OPP.) (Essentially just adding a bunch of if-statements to protect these fields)

**Review by SimonBaeumer — CHANGES_REQUESTED** [review]
The code looks good overall, but requesting changes because the `BuildProfileRefID` needs a short discussion on how to handle the fields as these know are cached in the `ScanRefID` UUID.

**Review by SimonBaeumer — APPROVED** [review]

**Review by SimonBaeumer — APPROVED** [review]
:ship:

**Comment by bradr5** [comment]
LGTM

Just want to point out that trying to get results for a single profile could result in multiple stats being returned due using profile name and prefix matching. i.e. ocp4-cis and ocp4-cis-node

`roxcurl "v2/compliance/scan/stats/profile?query=Compliance%20Profile%20Name:ocp4-cis" | json_pp`
^this would return 2 results

With that said, I don't _think_ the UI needs to get scan stats on a single profile

**Review by bradr5 — APPROVED** [review]

**Comment by bradr5** [comment]
> LGTM
> 
> Just want to point out that trying to get results for a single profile could result in multiple stats being returned due using profile name and prefix matching. i.e. ocp4-cis and ocp4-cis-node
> 
> `roxcurl "v2/compliance/scan/stats/profile?query=Compliance%20Profile%20Name:ocp4-cis" | json_pp` ^this would return 2 results
> 
> With that said, I don't _think_ the UI needs to get scan stats on a single profile

edit: Actually it will make getting compliance percentage on the single profile view a little awkward but we can technically just do the filtering on the frontend
Screenshot 2024-03-15 at 11 24 08 AM
